pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -923,48 +1061,44 @@ func (t *Test) runCommandFunc(nodes option.NodeListOption, cmd string) stepFunc
 
 // chooseUpgradePath returns a valid upgrade path for the test to
 // take. An upgrade path is a list of predecessor versions that can
-// be upgraded into eachother, ending at the current build version.
+// be upgraded into each other, ending at the current build version.
 // It uses the `predecessorFunc` field to compute the actual list of
 // predecessors. This function may also choose to skip releases when
 // supported. Special care is taken to avoid using releases that are
 // not available under a certain cluster architecture. Specifically,
 // ARM64 builds are only available on v22.2.0+.
 func (t *Test) chooseUpgradePath() ([]*clusterupgrade.Version, error) {
 	skipVersions := t.prng.Float64() < t.options.skipVersionProbability
-	numUpgrades := t.numUpgrades()
-	currentVersion := clusterupgrade.CurrentVersion()
 
+	// Checks whether the predecessor is feasible subject to these constraints:
+	//  - it is not older than the minimum supported ARM64 version
+	//	- it is not older than the minimum bootstrap version
 	isAvailable := func(v *clusterupgrade.Version) bool {
-		if t.clusterArch() != vm.ArchARM64 {
-			return true
+		if t.clusterArch() == vm.ArchARM64 && !v.AtLeast(minSupportedARM64Version) {
+			return false
 		}
-
-		return v.AtLeast(minSupportedARM64Version)
-	}
-	isOlderThanMinimumBootstrapVersion := func(v *clusterupgrade.Version) bool {
-		return t.options.minimumBootstrapVersion != nil && v.LessThan(t.options.minimumBootstrapVersion)
+		return t.options.minimumBootstrapVersion == nil || v.AtLeast(t.options.minimumBootstrapVersion)
 	}
-
-	forceSkipVersion := true
+	// Set to true after the first skip-version upgrade is chosen.
+	skipToCurrentChosen := false
 	// possiblePredecessorsFor returns a list of possible predecessors
 	// for the given release `v`. If skip-version is enabled and
 	// supported, this function will return both the immediate
 	// predecessor along with the predecessor's predecessor. This is the
 	// function to change in case the rules around what upgrades are
 	// possible in CRDB change.
 	possiblePredecessorsFor := func(v *clusterupgrade.Version) ([]*clusterupgrade.Version, error) {
-		pred, err := t.options.predecessorFunc(t.prng, v, t.options.minimumSupportedVersion, t.options.minimumBootstrapVersion)
+		pred, err := t.options.predecessorFunc(t.prng, v)
 		if err != nil {
 			return nil, err
 		}
-
-		if !isAvailable(pred) {
-			return nil, nil
+		pred, err = maybeClampMsbMsv(t.prng, pred, t.options.minimumBootstrapVersion, t.options.minimumSupportedVersion)
```

Comment:
This whole detour of refactoring was really triggered by a confusing `predecessor` API; it very much felt like we were mixing the concern of selecting a (random) predecessor with the concern of "clamping" or adjusting patch version. Hence, they are now separate.